### PR TITLE
[Cache] Memcached Adapter Docs and Cache Pool Docs Refactoring

### LINKED ISSUE #7256: [Cache] Document the MemcachedAdapter
This one was added by @robfrawley in symfony/symfony#20858.

### PR DESCRIPTION
**Pull request by robfrawley**
Documentation for symfony/symfony#20863 and symfony/symfony#20858. PR symfony/symfony#20863 is still being written/finalized, so this is a WIP following that PR at this time. Resolves #7256.

### CODE REVIEW

**Comment by robfrawley** [comment]
Test failure unrelated to this PR: https://travis-ci.org/symfony/symfony-docs/builds/184129616#L215

**Comment by robfrawley** [comment]
Assuming symfony/symfony#20863 is merged as-is, documentation is up-to-date. Refactored the adapter pool docs to move them into separate files, as the document became very, very long with the detailed Memcached adapter docs added.

**Comment by robfrawley** [comment]
Documentation PR submitted (assuming this is merged as-is) as symfony/symfony-docs/#7265: http://pr-7265-vfyqlyy-6qmocelev2lwe.eu.platform.sh/components/cache/adapters/memcached_adapter.html

**Comment by robfrawley** [comment]
Test failure unrelated: https://travis-ci.org/symfony/symfony/jobs/184824689#L3482

**Comment by nicolas-grekas** [comment]
Great work! Instead of listing the options, I think we should link to the php doc. And btw, if you could expand to the same level the redis section, it would be great to have them on parity.

**Comment by robfrawley** [comment]
Regarding Redis: I'll give that a look and update it in-line with Memcached adapter; sure.

As for the Memcached options, I'm split on this. The original copy didn't have options, and later they were added for a few reasons. First, the PHP manual includes invalid options, like `Memcached::OPT_CACHE_LOOKUPS` which will cause a deprecation exception to be thrown by the extension. Second, the PHP manual is missing options, like `randomize_replica_read`.  And third, the PHP manual doesn't do a great job of making it obvious what are option names and what are option values. Hence: the inclusion of options in the Symfony docs. @nicolas-grekas Given the above, do you still think we should just link to the PHP manual? It should be noted that I *still do* link to the manual as the last line of the documentation.

**Comment by nicolas-grekas** [comment]
Reasonable :) let's see what the docs core team thinks about it, maybe there is some existing practice on the topic.
I may have missed it, but we should tell that v2.2 or higher is required btw!

**Comment by robfrawley** [comment]
Good catch (re version requirement). Will add.

File: components/cache/adapters/memcached_adapter.rst
Comment (javiereguiluz) [inline components/cache/adapters/memcached_adapter.rst]: Here we're missing the default value of `weight`: 0? 50?

**Review by javiereguiluz — APPROVED** [review]
@robfrawley thanks for contributing this massive PR. I've reviewed it carefully and I'm 👍 

In 68dc310 I've made some minor simplifications and I've sorted the options alphabetically and updated its syntax to use the reStructuredText definition lists (that's why the changes look bigger than they really are).

**Comment by robfrawley** [comment]
@javiereguiluz Thanks for taking a look and reviewing this PR; I won't have time to update this PR until the coming weekend. You can expect me to address any comments you made, as well as expand the Redis section. I'll be sure to submit something before the end of the weekend, and then this should likely be merged ASAP, as all related PRs have been finalized and submitted into the Symfony codebase already. I'll let @javiereguiluz and @nicolas-grekas know when this needs a final review. Thanks!

File: components/cache/adapters/memcached_adapter.rst
Comment (robfrawley) [inline components/cache/adapters/memcached_adapter.rst]: So, there isn't really a "default" for this value. If a weight isn't set, all servers are treated exactly the same. We don't provide any default weight when one isn't specified. How do you think we should make that clear and word it?

**Comment by robfrawley** [comment]
@nicolas-grekas Can you review the Redis documentation? The DSN syntax is shown as 
```
redis://[user:pass@][ip|host|socket[:port]][/db-index]
```
Are there any additional parts to that? Also, can you provide the project names (so I can link to them) for the respective extensions and/or user-space class implementations that provide `Redis`, `RedisArray`, `RedisCluster`, and `Predis`. And lastly, how do you want to handle the options for this adapter?

**Comment by robfrawley** [comment]
@javiereguiluz Do you know why `platform.sh` is failing during build deployment?

**Comment by javiereguiluz** [comment]
@robfrawley yes ... but I don't know how to solve it. Platform.sh broke for us a few weeks ago ... but I don't know what to do. Maybe @weaverryan or @crell could take a look at our error trace:

```
Building application 'symfonydocs' (runtime type: php, tree: a1d132d)
  Generating runtime configuration.
  
  Installing build dependencies...
    Installing python build dependencies: sphinx
  
    E: Command exited with error code 2. Output follows:
      Downloading/unpacking sphinx==1.3.6 (from -r /app/.global/requirements.txt (line 1))
        Using download cache from /mnt/cache/pip/https%3A%2F%2Fpypi.python.org%2Fpackages%2Fea%2F56%2F9d723915960f50010f5d98874d9c357ff048661611e0b09dc726622c5c3d%2FSphinx-1.3.6.tar.gz
        Running setup.py egg_info for package sphinx
          
          warning: no files found matching 'TODO'
          no previously-included directories found matching 'doc/_build'
      Downloading/unpacking six>=1.4 (from sphinx==1.3.6->-r /app/.global/requirements.txt (line 1))
        Using download cache from /mnt/cache/pip/https%3A%2F%2Fpypi.python.org%2Fpackages%2Fb3%2Fb2%2F238e2590826bfdd113244a40d9d3eb26918bd798fc187e2360a8367068db%2Fsix-1.10.0.tar.gz
        Running setup.py egg_info for package six
          
          no previously-included directories found matching 'documentation/_build'
      Downloading/unpacking Jinja2>=2.3 (from sphinx==1.3.6->-r /app/.global/requirements.txt (line 1))
        Using download cache from /mnt/cache/pip/https%3A%2F%2Fpypi.python.org%2Fpackages%2Ff4%2F3f%2F28387a5bbc6883082c16784c6135440b94f9d5938fb156ff579798e18eda%2FJinja2-2.9.4.tar.gz
        Running setup.py egg_info for package Jinja2
          
          warning: no files found matching 'run-tests.py'
          warning: no files found matching '*' under directory 'custom_fixers'
          warning: no files found matching '*' under directory 'jinja2/testsuite/res'
          warning: no previously-included files matching '*' found under directory 'docs/_build'
          warning: no previously-included files matching '*.pyc' found under directory 'jinja2'
          warning: no previously-included files matching '*.pyc' found under directory 'docs'
          warning: no previously-included files matching '*.pyo' found under directory 'jinja2'
          warning: no previously-included files matching '*.pyo' found under directory 'docs'
      Downloading/unpacking Pygments>=2.0 (from sphinx==1.3.6->-r /app/.global/requirements.txt (line 1))
        Using download cache from /mnt/cache/pip/https%3A%2F%2Fpypi.python.org%2Fpackages%2Fb8%2F67%2Fab177979be1c81bc99c8d0592ef22d547e70bb4c6815c383286ed5dec504%2FPygments-2.1.3.tar.gz
        Running setup.py egg_info for package Pygments
          
      Downloading/unpacking docutils>=0.11 (from sphinx==1.3.6->-r /app/.global/requirements.txt (line 1))
        Using download cache from /mnt/cache/pip/https%3A%2F%2Fpypi.python.org%2Fpackages%2F05%2F25%2F7b5484aca5d46915493f1fd4ecb63c38c333bd32aa9ad6e19da8d08895ae%2Fdocutils-0.13.1.tar.gz
        Running setup.py egg_info for package docutils
          
          warning: no files found matching 'MANIFEST'
          warning: no previously-included files matching '*.pyc' found under directory '*'
          warning: no previously-included files matching '*~' found under directory '*'
          warning: no previously-included files matching '.DS_Store' found under directory '*'
      Downloading/unpacking snowballstemmer>=1.1 (from sphinx==1.3.6->-r /app/.global/requirements.txt (line 1))
        Using download cache from /mnt/cache/pip/https%3A%2F%2Fpypi.python.org%2Fpackages%2F20%2F6b%2Fd2a7cb176d4d664d94a6debf52cd8dbae1f7203c8e42426daa077051d59c%2Fsnowballstemmer-1.2.1.tar.gz
        Running setup.py egg_info for package snowballstemmer
          
          warning: no files found matching '*.py' under directory 'src'
      Downloading/unpacking babel>=1.3,!=2.0 (from sphinx==1.3.6->-r /app/.global/requirements.txt (line 1))
        Using download cache from /mnt/cache/pip/https%3A%2F%2Fpypi.python.org%2Fpackages%2F6e%2F96%2Fba2a2462ed25ca0e651fb7b66e7080f5315f91425a07ea5b34d7c870c114%2FBabel-2.3.4.tar.gz
        Running setup.py egg_info for package babel
          
          warning: no previously-included files matching '*' found under directory 'docs/_build'
          warning: no previously-included files matching '*.pyc' found under directory 'tests'
          warning: no previously-included files matching '*.pyo' found under directory 'tests'
      Downloading/unpacking alabaster>=0.7,<0.8 (from sphinx==1.3.6->-r /app/.global/requirements.txt (line 1))
        Using download cache from /mnt/cache/pip/https%3A%2F%2Fpypi.python.org%2Fpackages%2F71%2Fc3%2F70da7d8ac18a4f4c502887bd2549e05745fa403e2cd9d06a8a9910a762bc%2Falabaster-0.7.9.tar.gz
        Running setup.py egg_info for package alabaster
          
          warning: no files found matching 'README.md'
          warning: no previously-included files matching '.*.swp' found under directory 'alabaster'
      Downloading/unpacking sphinx-rtd-theme>=0.1,<2.0 (from sphinx==1.3.6->-r /app/.global/requirements.txt (line 1))
        Using download cache from /mnt/cache/pip/https%3A%2F%2Fpypi.python.org%2Fpackages%2Fda%2F6b%2F1b75f13d8aa3333f19c6cdf1f0bc9f52ea739cae464fbee050307c121857%2Fsphinx_rtd_theme-0.1.10a0.tar.gz
        Running setup.py egg_info for package sphinx-rtd-theme
          
      Downloading/unpacking MarkupSafe>=0.23 (from Jinja2>=2.3->sphinx==1.3.6->-r /app/.global/requirements.txt (line 1))
        Using download cache from /mnt/cache/pip/https%3A%2F%2Fpypi.python.org%2Fpackages%2Fc0%2F41%2Fbae1254e0396c0cc8cf1751cb7d9afc90a602353695af5952530482c963f%2FMarkupSafe-0.23.tar.gz
        Running setup.py egg_info for package MarkupSafe
          
      Downloading/unpacking pytz>=0a (from babel>=1.3,!=2.0->sphinx==1.3.6->-r /app/.global/requirements.txt (line 1))
        Using download cache from /mnt/cache/pip/https%3A%2F%2Fpypi.python.org%2Fpackages%2F42%2F00%2F5c89fc6c9b305df84def61863528e899e9dccb196f8438f6cbe960758fc5%2Fpytz-2016.10.tar.gz
        Running setup.py egg_info for package pytz
          
      Installing collected packages: sphinx, six, Jinja2, Pygments, docutils, snowballstemmer, babel, alabaster, sphinx-rtd-theme, MarkupSafe, pytz
        Running setup.py install for sphinx
          
          warning: no files found matching 'TODO'
          no previously-included directories found matching 'doc/_build'
          Installing sphinx-apidoc script to /app/.global/bin
          Installing sphinx-build script to /app/.global/bin
          Installing sphinx-quickstart script to /app/.global/bin
          Installing sphinx-autogen script to /app/.global/bin
        Running setup.py install for six
          
          no previously-included directories found matching 'documentation/_build'
        Running setup.py install for Jinja2
          
          warning: no files found matching 'run-tests.py'
          warning: no files found matching '*' under directory 'custom_fixers'
          warning: no files found matching '*' under directory 'jinja2/testsuite/res'
          warning: no previously-included files matching '*' found under directory 'docs/_build'
          warning: no previously-included files matching '*.pyc' found under directory 'jinja2'
          warning: no previously-included files matching '*.pyc' found under directory 'docs'
          warning: no previously-included files matching '*.pyo' found under directory 'jinja2'
          warning: no previously-included files matching '*.pyo' found under directory 'docs'
          SyntaxError: ('invalid syntax', ('/app/.global/lib/python2.7/site-packages/jinja2/asyncfilters.py', 7, 9, 'async def auto_to_seq(value):\n'))
          
          SyntaxError: ('invalid syntax', ('/app/.global/lib/python2.7/site-packages/jinja2/asyncsupport.py', 22, 9, 'async def concat_async(async_gen):\n'))
          
        Running setup.py install for Pygments
          
          Installing pygmentize script to /app/.global/bin
        Running setup.py install for docutils
          changing mode of build/scripts-2.7/rst2html.py from 644 to 755
          changing mode of build/scripts-2.7/rst2html5.py from 644 to 755
          changing mode of build/scripts-2.7/rst2s5.py from 644 to 755
          changing mode of build/scripts-2.7/rst2latex.py from 644 to 755
          changing mode of build/scripts-2.7/rst2xetex.py from 644 to 755
          changing mode of build/scripts-2.7/rst2man.py from 644 to 755
          changing mode of build/scripts-2.7/rst2xml.py from 644 to 755
          changing mode of build/scripts-2.7/rst2pseudoxml.py from 644 to 755
          changing mode of build/scripts-2.7/rstpep2html.py from 644 to 755
          changing mode of build/scripts-2.7/rst2odt.py from 644 to 755
          changing mode of build/scripts-2.7/rst2odt_prepstyles.py from 644 to 755
          
          warning: no files found matching 'MANIFEST'
          warning: no previously-included files matching '*.pyc' found under directory '*'
          warning: no previously-included files matching '*~' found under directory '*'
          warning: no previously-included files matching '.DS_Store' found under directory '*'
          changing mode of /app/.global/bin/rst2xetex.py to 755
          changing mode of /app/.global/bin/rst2html.py to 755
          changing mode of /app/.global/bin/rst2odt.py to 755
          changing mode of /app/.global/bin/rst2odt_prepstyles.py to 755
          changing mode of /app/.global/bin/rst2html5.py to 755
          changing mode of /app/.global/bin/rst2pseudoxml.py to 755
          changing mode of /app/.global/bin/rst2s5.py to 755
          changing mode of /app/.global/bin/rst2xml.py to 755
          changing mode of /app/.global/bin/rstpep2html.py to 755
          changing mode of /app/.global/bin/rst2man.py to 755
          changing mode of /app/.global/bin/rst2latex.py to 755
        Running setup.py install for snowballstemmer
          
          warning: no files found matching '*.py' under directory 'src'
        Running setup.py install for babel
          
          warning: no previously-included files matching '*' found under directory 'docs/_build'
          warning: no previously-included files matching '*.pyc' found under directory 'tests'
          warning: no previously-included files matching '*.pyo' found under directory 'tests'
          Installing pybabel script to /app/.global/bin
        Running setup.py install for alabaster
          
          warning: no files found matching 'README.md'
          warning: no previously-included files matching '.*.swp' found under directory 'alabaster'
        Running setup.py install for sphinx-rtd-theme
          
        Found existing installation: MarkupSafe 0.15
          Uninstalling MarkupSafe:
      Exception:
      Traceback (most recent call last):
        File "/usr/lib/python2.7/dist-packages/pip/basecommand.py", line 104, in main
          status = self.run(options, args)
        File "/usr/lib/python2.7/dist-packages/pip/commands/install.py", line 250, in run
          requirement_set.install(install_options, global_options)
        File "/usr/lib/python2.7/dist-packages/pip/req.py", line 1129, in install
          requirement.uninstall(auto_confirm=True)
        File "/usr/lib/python2.7/dist-packages/pip/req.py", line 486, in uninstall
          paths_to_remove.remove(auto_confirm)
        File "/usr/lib/python2.7/dist-packages/pip/req.py", line 1431, in remove
          renames(path, new_path)
        File "/usr/lib/python2.7/dist-packages/pip/util.py", line 263, in renames
          shutil.move(old, new)
        File "/usr/lib/python2.7/shutil.py", line 299, in move
          rmtree(src)
        File "/usr/lib/python2.7/shutil.py", line 252, in rmtree
          onerror(os.remove, fullname, sys.exc_info())
        File "/usr/lib/python2.7/shutil.py", line 250, in rmtree
          os.remove(fullname)
      OSError: [Errno 30] Read-only file system: '/usr/lib/python2.7/dist-packages/MarkupSafe-0.15.egg-info/PKG-INFO'
      
      Storing complete log in /app/.pip/pip.log
  
  E: Error building project: `pip` could not be run. Aborted build.

E: Error building the project: Unable to build project, aborting deployment.
```

**Comment by robfrawley** [comment]
@javiereguiluz The error looks like it is related to pypa/pip#1818: a system-level package is trying to be uninstalled to resolve user-level (non-sudo) installation. What version of `pip` is being used? According to https://github.com/pypa/pip/issues/1818#issuecomment-149720643 it has been fixed in 1.1 and 1.5 (but not 1.4) by passing the `--ignore-installed` option to `pip`.

File: components/cache/adapters/chain_adapter.rst
Comment (nicolas-grekas) [inline components/cache/adapters/chain_adapter.rst]: this is a simple way of creating a layered cache, from fastest to slowest storages::

File: components/cache/adapters/pdo_doctrine_dbal_adapter.rst
Comment (nicolas-grekas) [inline components/cache/adapters/pdo_doctrine_dbal_adapter.rst]: **in** a SQL

File: components/cache/adapters/redis_adapter.rst
Comment (nicolas-grekas) [inline components/cache/adapters/redis_adapter.rst]: RedisAdapter::clear()

File: components/cache/adapters/redis_adapter.rst
Comment (nicolas-grekas) [inline components/cache/adapters/redis_adapter.rst]: This variant is not implemented yet, so maybe not document it for now?

File: components/cache/adapters/redis_adapter.rst
Comment (nicolas-grekas) [inline components/cache/adapters/redis_adapter.rst]: "array" variant not implemented

File: components/cache/adapters/redis_adapter.rst
Comment (nicolas-grekas) [inline components/cache/adapters/redis_adapter.rst]: missing link to some doc listing all available options?

File: components/cache/adapters/redis_adapter.rst
Comment (nicolas-grekas) [inline components/cache/adapters/redis_adapter.rst]: https://github.com/phpredis/phpredis/blob/master/arrays.markdown#readme

File: components/cache/adapters/redis_adapter.rst
Comment (nicolas-grekas) [inline components/cache/adapters/redis_adapter.rst]: https://github.com/phpredis/phpredis/blob/master/cluster.markdown#readme

**Review by nicolas-grekas — COMMENTED** [review]
nice

**Comment by robfrawley** [comment]
@javiereguiluz Any luck following the advice in https://github.com/symfony/symfony-docs/pull/7265#issuecomment-273833614?

File: components/cache/adapters/apcu_adapter.rst
Comment (robfrawley) [inline components/cache/adapters/apcu_adapter.rst]: Done

File: components/cache/adapters/apcu_adapter.rst
Comment (robfrawley) [inline components/cache/adapters/apcu_adapter.rst]: Done, and cleaned up entire APCu docs a bit.

File: components/cache/adapters/array_cache_adapter.rst
Comment (robfrawley) [inline components/cache/adapters/array_cache_adapter.rst]: Used your re-write to expand a bit more, to:
```
Generally, this adapter is useful for testing purposes, as its contents are stored in memory
and not persisted outside the running PHP process in any way. It can also be useful while
warming up caches, due to the :method:`Symfony\\Component\\Cache\\Adapter\\ArrayAdapter::getValues()`
method.

This adapter can be passed a default cache lifetime as its first parameter, and a serialization
toggle as its second parameter::
```

File: components/cache/adapters/chain_adapter.rst
Comment (robfrawley) [inline components/cache/adapters/chain_adapter.rst]: Using `so this is a simple way of creating a layered cache, from fastest to slowest storage engines::`

**Comment by javiereguiluz** [comment]
@robfrawley sadly I don't know where to make any change. The only files related to Platform.sh are stored here: https://github.com/symfony/symfony-docs/tree/master/.platform

File: components/cache/adapters/pdo_doctrine_dbal_adapter.rst
Comment (robfrawley) [inline components/cache/adapters/pdo_doctrine_dbal_adapter.rst]: done

File: components/cache/adapters/redis_adapter.rst
Comment (robfrawley) [inline components/cache/adapters/redis_adapter.rst]: Do we want to get a PR together for it in the immediate so as to offer the same ability for both MemcachedAdapter and RedisAdapter? If you don't have the time, I can look to getting a PR together? Or do you not care one way or the other? How much time do we have before the 3.3 release (when these docs need to be live)?

File: components/cache/adapters/redis_adapter.rst
Comment (robfrawley) [inline components/cache/adapters/redis_adapter.rst]: :+1:

File: components/cache/adapters/redis_adapter.rst
Comment (robfrawley) [inline components/cache/adapters/redis_adapter.rst]: :+1:

File: components/cache/adapters/redis_adapter.rst
Comment (robfrawley) [inline components/cache/adapters/redis_adapter.rst]: :+1:

File: components/cache/adapters/redis_adapter.rst
Comment (robfrawley) [inline components/cache/adapters/redis_adapter.rst]: Ditto: https://github.com/symfony/symfony-docs/pull/7265#discussion_r97227095

File: components/cache/adapters/redis_adapter.rst
Comment (nicolas-grekas) [inline components/cache/adapters/redis_adapter.rst]: 3.3 is feature freeze by the end of march.
I won't be able to work on this until then.
If you'd like to do it, just note that when an array is passed, we have to handle both phpredis & Predis, and both array & cluster modes.
See https://github.com/nrk/predis#aggregate-connections for Predis
The "class" option should be the way to go from the userland side.

File: components/cache/adapters/redis_adapter.rst
Comment (robfrawley) [inline components/cache/adapters/redis_adapter.rst]: Yeah, how do we want to handle the options for Redis, since it supports a few extension/user-land class implementations?

**Comment by robfrawley** [comment]
@javiereguiluz @nicolas-grekas Big update just committed, which requires @javiereguiluz's proof-reading skills once again. If you can take a look at the latest changes as well @nicolas-grekas, that would be great (I think I addressed all your concersns, as well as standardized the docs for all adapters a bit). Most-recent diff: https://github.com/symfony/symfony-docs/pull/7265/commits/bbb7d1a1da5a70dc4c7d97f5c5ac85741278febd

File: components/cache/adapters/memcached_adapter.rst
Comment (nicolas-grekas) [inline components/cache/adapters/memcached_adapter.rst]: APCu

File: components/cache/adapters/memcached_adapter.rst
Comment (nicolas-grekas) [inline components/cache/adapters/memcached_adapter.rst]: the amount

File: components/cache/adapters/redis_adapter.rst
Comment (nicolas-grekas) [inline components/cache/adapters/redis_adapter.rst]: APCu

File: components/cache/adapters/redis_adapter.rst
Comment (nicolas-grekas) [inline components/cache/adapters/redis_adapter.rst]: Dunno, please choose what's most sensible to you.

File: components/cache/adapters/memcached_adapter.rst
Comment (robfrawley) [inline components/cache/adapters/memcached_adapter.rst]: done

File: components/cache/adapters/memcached_adapter.rst
Comment (robfrawley) [inline components/cache/adapters/memcached_adapter.rst]: done

File: components/cache/adapters/redis_adapter.rst
Comment (robfrawley) [inline components/cache/adapters/redis_adapter.rst]: done

**Comment by robfrawley** [comment]
@javiereguiluz and @nicolas-grekas Including commit from #7402 for now so the output can be seen for easier review.

**Comment by nicolas-grekas** [comment]
:+1: on my side - would be great to merge soon so that we can start working on the doc for PSR-16 (see https://github.com/symfony/symfony/pull/20694)

**Comment by robfrawley** [comment]
@nicolas-grekas Last question before this should be set for merge: are all of the following allowed options (depending on client, of course)?
- https://github.com/phpredis/phpredis/blob/develop/redis.c#L533-L555
- https://github.com/nrk/predis/wiki/Connection-Parameters#list-of-connection-parameters
- https://github.com/phpredis/phpredis/blob/master/arrays.markdown#specifying-the-retry_interval-parameter

Am I missing anything?

**Comment by nicolas-grekas** [comment]
@robfrawley LGTM yes, with those listed at https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Cache/Traits/RedisTrait.php#L27

**Comment by nicolas-grekas** [comment]
Oh wait, in fatc for phpredis, we we don't support any other than the listed ones in $defaultConnectionOptions - eg no config for the serializer.
For Predis, we support all of them since they are all given to the Predis connectin factory.

File: components/cache/adapters/redis_adapter.rst
Comment (robfrawley) [inline components/cache/adapters/redis_adapter.rst]: Done. Added the following:

```rst
Available Options
~~~~~~~~~~~~~~~~~

``class`` (type: ``string``)
    Specifies the connection library to return, either ``\Redis`` or ``\Predis\Client``.
    If none is specified, it will return ``\Redis`` if the ``redis`` extension is
    available, and ``\Predis\Client`` otherwise.

``persistent`` (type: ``int``, default: ``0``)
    Enables or disables use of persistent connections. A value of ``0`` disables persistent
    connections, and a value of ``1`` enables them.

``persistent_id`` (type: ``string|null``, default: ``null``)
    Specifies the persistent id string to use for a persistent connection.

``read_timeout`` (type: ``int``, default: ``0``)
    Specifies the time (in seconds) used when performing read operations on the underlying
    network resource before the operation times out.

``retry_interval`` (type: ``int``, default: ``0``)
    Specifies the delay (in milliseconds) between reconnection attempts in case the client
    loses connection with the server.

``timeout`` (type: ``int``, default: ``30``)
    Specifies the time (in seconds) used to connect to a Redis server before the
    connection attempt times out.

.. note::
    When using the `Predis`_ library some additional Predis-specific options are available.
    Reference the `Predis Connection Parameters`_ documentation for more information.
```

**Comment by robfrawley** [comment]
@nicolas-grekas Finished the section pertaining to Redis connection options (see [the platform.sh build](http://pr-7265-vfyqlyy-6qmocelev2lwe.eu.platform.sh/components/cache/adapters/redis_adapter.html#available-options)). Feel free to advise me if you require additional changes.

@javiereguiluz Did some rebasing to clean this up a bit; let me know if I should squash or otherwise edit the commit history prior to the PR's merger.

This should be ready!

**Comment by nicolas-grekas** [comment]
Doc what the default "persistent_id" is when only persistent is set? (I don't remember myself)

**Comment by robfrawley** [comment]
@nicolas-grekas Didn't look like that was the case when I gave it a once-over: https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Cache/Traits/RedisTrait.php#L105

Please correct me if I am wrong and should add something about `persistend_id` only applying when `persistend` is enabled, but it doesn't look like that is the case to me.

**Comment by robfrawley** [comment]
Erm, nevermind, misread your comment. The default *is* `null`.

That said, the behavior looks different for `\Redis` and `\Predis\Client`: with `\Redis` it will enable persistent connections if either `persistent` or `persistent_id` is set, but with `Predis\Client` it just blindly passes the parameters. Does that client behave similarly?

**Comment by robfrawley** [comment]
Is this good to go, or do I need to do anything prior to merge?

**Comment by xabbuh** [comment]
@robfrawley Can you resolve conflicts here?

**Comment by robfrawley** [comment]
@xabbuh rebased

File: components/cache/adapters/apcu_adapter.rst
Comment (xabbuh) [inline components/cache/adapters/apcu_adapter.rst]: filesystem

File: components/cache/adapters/apcu_adapter.rst
Comment (xabbuh) [inline components/cache/adapters/apcu_adapter.rst]: adapter's

File: components/cache/adapters/chain_adapter.rst
Comment (xabbuh) [inline components/cache/adapters/chain_adapter.rst]: The following example [...]

**Comment by robfrawley** [comment]
@xabbuh fixes pushed

**Comment by xabbuh** [comment]
:+1:

Status: Reviewed

**Comment by robfrawley** [comment]
@javiereguiluz @xabbuh Anything else you guys want me to do for this PR? This should likely be merged sooner than later, as (aside from the general cleanup and expansion of adapter docs) provides initial documentation for the Memcached feature, which has been live on `master` for over a month. Additionally, as this moves adapter onto their own pages, it will continue to have merge conflicts when anyone touches adapter docs, and I'd appreciate not having to rebase over and over. ;-) All the best.

**Comment by javiereguiluz** [comment]
@xabbuh nothing else from me ... just thanking @robfrawley again for his amazing work here!

**Comment by robfrawley** [comment]
Thanks for the reviews guys, all for the love of Symfony. ;-)

**Comment by xabbuh** [comment]
Thank you @robfrawley.

**Pull request merged**
